packages/btcindexer/src/btcindexer.ts

```diff
@@ -144,6 +151,81 @@ export class Indexer {
 		return deposits;
 	}
 
+	async processFinalizedTransactions(): Promise<void> {
+		const finalizedTxs = await this.d1
+			.prepare("SELECT tx_id, block_hash FROM nbtc_txs WHERE status = 'finalized'")
+			.all<{ tx_id: string; block_hash: string }>();
+
+		if (!finalizedTxs.results || finalizedTxs.results.length === 0) {
+			return;
+		}
+
+		const updates: D1PreparedStatement[] = [];
+		// TODO: do we imidietly process it and check if it was succesful and just change the status to minted? This should be a matter of seconds at most.
+		const setMintingStmt = this.d1.prepare(
+			"UPDATE nbtc_txs SET status = 'minting', updated_at = CURRENT_TIMESTAMP WHERE tx_id = ?",
+		);
+
+		for (const txInfo of finalizedTxs.results) {
+			try {
+				const rawBlockBuffer = await this.blocksDB.get(txInfo.block_hash, {
+					type: "arrayBuffer",
+				});
+				if (!rawBlockBuffer) {
+					continue;
+				}
+
+				const block = Block.fromBuffer(Buffer.from(rawBlockBuffer));
+				const txIndex = block.transactions?.findIndex(
+					(tx) => tx.getId() === txInfo.tx_id,
+				);
+				const targetTx = block.transactions?.[txIndex ?? -1];
+
+				if (txIndex === undefined || txIndex === -1 || !targetTx) {
+					continue;
+				}
+				const proof = this.constructMerkleProof(block, targetTx);
+				if (!proof) {
+					continue;
+				}
+
+				// soundness check
+				if (proof.merkleRoot !== block.merkleRoot?.toString("hex")) {
+					continue;
+				}
+
+				// TODO: Call the minting smart contract.
+				// await suiClient.mintNBTC();
+				updates.push(setMintingStmt.bind(txInfo.tx_id));
+			} catch (e) {
+				console.error(`Failed to process finalized tx ${txInfo.tx_id}:`, e);
+			}
+		}
+
+		if (updates.length > 0) {
+			try {
+				await this.d1.batch(updates);
+			} catch (e) {
+				console.error(`failed to apply batch updates to D1.`, e);
+			}
+		}
+	}
+
+	constructMerkleProof(block: Block, targetTx: Transaction): ProofResult | null {
+		if (!block.transactions || block.transactions.length === 0) {
+			return null;
+		}
+
+		const leaves = block.transactions.map((tx) => Buffer.from(tx.getHash()).reverse());
+		const targetLeaf = Buffer.from(targetTx.getHash()).reverse();
+		const tree = new MerkleTree(leaves, SHA256, { isBitcoinTree: true });
```

Comment:
I think @vuvoth meant to cache this part. Let's break down the proving as follows:
```
const merkleTree = constructMerkleTree(block);
for (const tx of nbtc_txs) {
   const proof = constructTxProof(merkleTree, tx)
   ...
}
```

Reply:
done